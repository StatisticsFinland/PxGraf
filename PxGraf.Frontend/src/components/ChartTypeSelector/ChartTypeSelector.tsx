import { useTranslation } from 'react-i18next';
import { ToggleButton, ToggleButtonGroup } from '@mui/material';
import React from 'react';

interface IChartTypeSelectorProps {
    possibleTypes: string[];
    selectedType?: string;
    onTypeSelected: (value: string) => void;
}

export const ChartTypeSelector: React.FC<IChartTypeSelectorProps> = ({ possibleTypes, selectedType, onTypeSelected }) => {
    const { t } = useTranslation();
    const isSelected = (type) => {
        return !selectedType ? type === possibleTypes[0] : selectedType === type;
    };

    return (
        <ToggleButtonGroup
            aria-label={t('tooltip.visualizationType')}
            value={"1"}
            exclusive
        >
            {!possibleTypes || possibleTypes.length < 1
                ? <span>{t("general.noPossibleVisualizations")}</span>
                : possibleTypes.map((type) => {
                    return <ToggleButton value={type}
                        key={type}
                        onClick={(_event, value) => onTypeSelected(value)}
                        selected={isSelected(type)}
                    >
                        {isSelected(type) ? <b>{t('chartTypes.' + type)}</b> : t('chartTypes.' + type)}
                    </ToggleButton>
                })
            }
        </ToggleButtonGroup>
    );
}

export default ChartTypeSelector;