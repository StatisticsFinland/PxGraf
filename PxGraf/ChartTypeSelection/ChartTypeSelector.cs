using Px.Utils.Models.Data.DataValue;
using Px.Utils.Models;
using PxGraf.ChartTypeSelection.ChartSpecificLimits;
using PxGraf.ChartTypeSelection.JsonObjects;
using PxGraf.Enums;
using PxGraf.Models.Queries;
using PxGraf.Models.SavedQueries;
using System.Collections.Generic;

namespace PxGraf.ChartTypeSelection
{
    /// <summary>
    /// Determines which kind of charts can be drawn from the data. The limits for each chart type are read from json.
    /// </summary>
    public class ChartTypeSelector : IChartTypeSelector
    {
        /// <summary>
        /// Provides the selector instance.
        /// </summary>
        public static IChartTypeSelector Selector
        {
            get
            {
                _selector ??= new ChartTypeSelector(new ChartSelectionLimits());
                return _selector;
            }
        }

        private static IChartTypeSelector _selector = null;

        /// <summary>
        /// Validity information about basic vertical bar chart with the given query
        /// </summary>
        private VerticalBarChartCheck VerticalBarChart { get; }

        /// <summary>
        /// Validity information about group vertical bar chart with the given query
        /// </summary>
        private GroupVerticalBarChartCheck GroupVerticalBarChart { get; }

        /// <summary>
        /// Validity information about stacked vertical bar chart with the given query
        /// </summary>
        private StackedVerticalBarChartCheck StackedVerticalBarChart { get; }

        /// <summary>
        /// Validity information about percent vertical bar chart with the given query.
        /// </summary>
        private PercentVerticalBarChartCheck PercentVerticalBarChart { get; }

        /// <summary>
        /// Validity information about basic horizontal bar chart with the given query
        /// </summary>
        private HorizontalBarChartCheck HorizontalBarChart { get; }

        /// <summary>
        /// Validity information about group horizontal bar chart with the given query
        /// </summary>
        private GroupHorizontalBarChartCheck GroupHorizontalBarChart { get; }

        /// <summary>
        /// Validity information about stacked horizontal bar chart with the given query
        /// </summary>
        private StackedHorizontalBarChartCheck StackedHorizontalBarChart { get; }

        /// <summary>
        /// Validity information about percent horizontal bar chart with the given query
        /// </summary>
        private PercentHorizontalBarChartCheck PercentHorizontalBarChart { get; }

        /// <summary>
        /// Validity information about pie chart with the given query
        /// </summary>
        private PieChartCheck PieChart { get; }

        /// <summary>
        /// Validity information about line chart with the given query
        /// </summary>
        private LineChartCheck LineChart { get; }

        /// <summary>
        /// Validity information about scatter plot with the given query
        /// </summary>
        private ScatterPlotCheck ScatterPlot { get; }

        /// <summary>
        /// Validity information about pyramid chart with the given query
        /// </summary>
        private PyramidChartCheck PyramidChart { get; }

        /// <summary>
        /// Validity information about tables with the given query
        /// </summary>
        private TableCheck Table { get; }

        /// <summary>
        /// Validity information about all possible chart types
        /// </summary>
        private List<ChartRulesCheck> AllLimitChecks { get; }

        /// <summary>
        /// Construct the type selector with a set of limit objects and a logger
        /// </summary>
        /// <param name="limits"></param>
        private ChartTypeSelector(IChartSelectionLimits limits)
        {
            VerticalBarChart = new VerticalBarChartCheck(limits.VerticalBarChartLimits);
            GroupVerticalBarChart = new GroupVerticalBarChartCheck(limits.GroupVerticalBarChartLimits);
            StackedVerticalBarChart = new StackedVerticalBarChartCheck(limits.StackedVerticalBarChartLimits);
            PercentVerticalBarChart = new PercentVerticalBarChartCheck(limits.StackedVerticalBarChartLimits); // Percent vertical should have its own rule set

            HorizontalBarChart = new HorizontalBarChartCheck(limits.HorizontalBarChartLimits);
            GroupHorizontalBarChart = new GroupHorizontalBarChartCheck(limits.GroupHorizontalBarChartLimits);
            StackedHorizontalBarChart = new StackedHorizontalBarChartCheck(limits.StackedHorizontalBarChartLimits);
            PercentHorizontalBarChart = new PercentHorizontalBarChartCheck(limits.StackedHorizontalBarChartLimits); // Percent horizontal should have its own rule set

            PieChart = new PieChartCheck(limits.PieChartLimits);
            LineChart = new LineChartCheck(limits.LineChartLimits);
            ScatterPlot = new ScatterPlotCheck(limits.ScatterPlotLimits);
            PyramidChart = new PyramidChartCheck(limits.PyramidChartLimits);

            Table = new TableCheck(limits.TableLimits);

            AllLimitChecks =
            [
                LineChart,
                HorizontalBarChart,
                GroupHorizontalBarChart,
                StackedHorizontalBarChart,
                PercentHorizontalBarChart,
                VerticalBarChart,
                GroupVerticalBarChart,
                StackedVerticalBarChart,
                PercentVerticalBarChart,
                PieChart,
                PyramidChart,
                ScatterPlot,
                Table
            ];
        }

        public IReadOnlyList<VisualizationType> GetValidChartTypes(MatrixQuery query, Matrix<DecimalDataValue> matrix)
        {
            List<VisualizationType> validTypes = [];
            foreach (ChartRulesCheck check in AllLimitChecks)
            {
                if (check.CheckValidity(VisualizationTypeSelectionObject.FromQueryAndMatrix(query, matrix)).Count == 0)
                {
                    validTypes.Add(check.Type);
                }
            }

            return validTypes;
        }

        public IReadOnlyDictionary<VisualizationType, IReadOnlyList<ChartRejectionInfo>> GetRejectionReasons(MatrixQuery query, Matrix<DecimalDataValue> matrix)
        {
            Dictionary<VisualizationType, IReadOnlyList<ChartRejectionInfo>> rejectionReasons = [];
            foreach (ChartRulesCheck check in AllLimitChecks)
            {
                rejectionReasons[check.Type] = check.CheckValidity(VisualizationTypeSelectionObject.FromQueryAndMatrix(query, matrix));
            }

            return rejectionReasons;
        }
    }
}
