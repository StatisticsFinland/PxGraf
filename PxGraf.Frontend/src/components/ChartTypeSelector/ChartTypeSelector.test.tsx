import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { ChartTypeSelector } from './ChartTypeSelector';
import '@testing-library/jest-dom';

jest.mock('react-i18next', () => ({
    ...jest.requireActual('react-i18next'),
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => null),
            },
        };
    },
}));

const onTypeSelectedMock = jest.fn((value: string) => value);

const mockTypes = ['foo', 'bar', 'baz'];

describe('Rendering test', () => {
    it('renders correctly', () => {
        const { asFragment } = render(<ChartTypeSelector onTypeSelected={onTypeSelectedMock} possibleTypes={mockTypes} selectedType={'foo'} />);
        expect(asFragment()).toMatchSnapshot();
    });
});

describe('Assertion tests', () => {
    it('should invoke handler if clicked', () => {
        render(<ChartTypeSelector onTypeSelected={onTypeSelectedMock} possibleTypes={mockTypes} selectedType={'foo'} />);
        fireEvent.click(screen.getByText('chartTypes.bar'));
        expect(onTypeSelectedMock).toBeCalledTimes(1);
    });
});