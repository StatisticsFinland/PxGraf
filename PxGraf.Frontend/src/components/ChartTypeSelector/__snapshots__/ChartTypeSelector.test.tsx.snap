// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering test renders correctly 1`] = `
<DocumentFragment>
  <div
    aria-label="tooltip.visualizationType"
    class="MuiToggleButtonGroup-root MuiToggleButtonGroup-horizontal css-1qb8o-MuiToggleButtonGroup-root"
    role="group"
  >
    <button
      aria-pressed="true"
      class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-firstButton css-gn0oko-MuiButtonBase-root-MuiToggleButton-root"
      tabindex="0"
      type="button"
      value="foo"
    >
      <b>
        chartTypes.foo
      </b>
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-middleButton css-gn0oko-MuiButtonBase-root-MuiToggleButton-root"
      tabindex="0"
      type="button"
      value="bar"
    >
      chartTypes.bar
    </button>
    <button
      aria-pressed="false"
      class="MuiButtonBase-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-standard MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButtonGroup-lastButton css-gn0oko-MuiButtonBase-root-MuiToggleButton-root"
      tabindex="0"
      type="button"
      value="baz"
    >
      chartTypes.baz
    </button>
  </div>
</DocumentFragment>
`;
